fix(registry): configure registry url independent of asset url (#1300)

gongfarmer opened the pull request:
fixes #1300 which is incorrectly marked closed.

Adds a configuration key for the registry url.
Also, will always download the registry if the registry configuration is pinned to "latest".

This allows  mason to work in a secure corporate environment where the mason registry api is not available, but a mirror of `mason-org/mason-registry` is available.

Supports this neovim/mason configuration:

```lua
require('mason').setup({
  -- use curl / wget to resolve metadata, skips version check against https://api.mason-registry.dev
  providers = { "mason.providers.client" },

  registries = { "github:mason-org/mason-registry@latest" },

  github = {
    download_url_template = 'https://proxy.corporate.com/github/%s/releases/download/%s/%s',
    registry_url_template = 'https://proxy.corporate.com/github/%s/releases/%s/download/%s',
  }
})
```

I have been running neovim with this code change applied to mason for a few days now.
Testing:

Using custom configuration:
* verified that the timestamp is getting updated in `~/.local/share/nvim/mason/registries/github/mason-org/mason-registry/info.json` after repeated runs.
* enabled debug-level logging and verified that there are no errors and that the registry download proceeds normally

Using normal configuration:
* verified that nothing is broken by this change.

williamboman reviewed (commented):
Hey! I appreciate the PR! I'm however not sure this is how I'd like to solve this. There are two apparent issues:

1) The `registry_url_template` and `download_url_template` are the same setting. If I understand this correctly, this was added to enable a workaround where you can pin the registry to `latest`, to leverage a different API endpoint for this.
2) `latest` as a version is an immutable version and is not something I've designed around. It could possibly introduce weird behavior (although I don't think so) but more importantly complicate future work. Setting `latest` will also not work with the default `registry_url_template`, so in order to pin it to `latest` (which is already the default behavior) you need to manually know which URL to override `registry_url_template` with, that introduces too much fraction.

I think the best way for you to work around being behind a very closed off network would be to implement a custom "provider" and use that to resolve the latest version. [Here's the default, primary, provider implementation](https://github.com/williamboman/mason.nvim/blob/main/lua/mason/providers/registry-api/init.lua#L6-L8).  (note: the [provider interface ](https://github.com/williamboman/mason.nvim/blob/main/lua/mason-core/providers/init.lua#L47-L54)is not yet considered part of the public domain so it's prone to unannounced breaking changes)

williamboman commented:
So this can be done like this:

```lua
-- ~/.config/nvim/lua/mason-x/provider.lua
local fetch = require "mason-core.fetch"

---@type Provider
return {
    github = {
        get_latest_release = function(repo)
            return fetch(("https://proxy.corporate.com/api/github/%s/releases/latest"):format(repo))
        end,
    }
}
```

```lua
require("mason").setup {
  providers = {
    "mason-x.provider", -- this is the lua module name i.e. what you'd pass to require() it
    "mason.providers.client",
    "mason.providers.registry-api",
  }
}
```

williamboman commented:
You can probably also leverage [`gh-cli`](https://github.com/cli/cli/issues/2037) and configure it to [proxy requests](https://github.com/cli/cli/issues/2037). In which case you'd want to use the `mason.providers.client` provider.

gongfarmer commented:
Thank you for the feedback and for taking the time to look at my PR.

I will look into the provider option you provided, that looks achievable.

I am trying to support a team of devs behind my company's big firewall, several of whom are interested in trying neovim but are not knowledgeable in lua or nvim.  I've built a "standard" `~/.config/nvim` repo that they can clone which configures lazy and mason to use our internal repo mirror.   
Unfortunately this Mason issue adds some extra steps as I will have to inject some code into ~/.local/share/nvim/lazy/mason.nvim/ to add my custom provider implementation.

I can make this work.  I see that Mason is still a work in progress with big changes still occurring for some major pieces, such as changes to how registries work earlier this year.  
I hope that at some point it will be possible to configure custom urls for the registry and packages so that people in secure environments can use this cool software without needing to add custom code.

williamboman commented:
The Lua module needs to exist somewhere in `:h 'rtp'` in order to be loadable, so it doesn't necessarily need to be in `~/.local/share/nvim/lazy/mason.nvim/`. It can be placed somewhere in for example:

- ~/.config/nvim
- ~/.local/share/nvim/site
- /usr/local/share/nvim/site
- /usr/share/nvim/site

Technically, everything is a plugin if you think about it, including your personal configuration, so you could also host it under a GitHub repo and include it as an additional plugin. Your plugin manager (lazy.nvim et al) will/should add the location in which they clone repos to `'rtp'`.

gongfarmer commented:
> So this can be done like this:
> 
> ```lua
> -- ~/.config/nvim/lua/mason-x/provider.lua
> local fetch = require "mason-core.fetch"
> 
> ---@type Provider
> return {
>     github = {
>         get_latest_release = function(repo)
>             return fetch(("https://proxy.corporate.com/api/github/%s/releases/latest"):format(repo))
>         end,
>     }
> }
> ```
> 
>  ```lua
> require("mason").setup {
>   providers = {
>     "mason-x.provider", -- this is the lua module name i.e. what you'd pass to require() it
>     "mason.providers.client",
>     "mason.providers.registry-api",
>   }
> }
> ```

This is a really powerful idea, thank you.I'm trying to implement it now.

The example code you provided attempts to download release json from https://proxy.corporate.com/api/github/%s/releases/latest, but I don't have any copies of the real release json available.  That's only available from the API.

I modified your code to construct a release json which provides working URLs for checksums.txt and registry.json.zip.

Unfortunately Mason ignores the urls I am providing and instead tries to fetch registry.json.zip from a url based on the configured `github.download_url_template` (As seen here: https://github.com/williamboman/mason.nvim/blob/f7f81ab41b153e2902ebded401a8a0a6abe28607/lua/mason-registry/sources/github.lua#L149 )
My code is returning good registry json, but `registry/sources/github.lua` only gets the release name from there, and then constructs its own url.  Unfortunately I don't have a "real" release name at that point, all I have is working urls for the release assets.


My code:
```lua
-- ~/.config/nvim/lua/mason-x/provider.lua

-- Want to download this:
--     https://proxy.repo.corporate.com/generic-github-releases/mason-org/mason-registry/releases/latest/download/checksums.txt
--     https://proxy.repo.corporate.com/generic-github-releases/mason-org/mason-registry/releases/latest/download/registry.json.zip
local registry_url_template = 'https://proxy.repo.corporate.com/generic-github-releases/%s/releases/latest/download'
local log = require "mason-core.log"
local Result = require "mason-core.result"
local result_template_json=[=[
  {
    "id": 0,
    "tag_name": "latest",
    "draft": false,
    "prerelease": false,
    "assets": [
      {
        "id": 0,
        "name": "checksums.txt",
        "url": "%s/checksums.txt",
        "browser_download_url": "%s/checksums.txt",
        "created_at": "2023-06-01T00:00:00Z",
        "updated_at": "2023-06-01T00:00:00Z",
        "size": 164,
        "download_count": 4
      },
      {
        "id": 0,
        "name": "registry.json.zip",
        "url": "%s/registry.json.zip",
        "browser_download_url": "%s/registry.json.zip",
        "created_at": "2023-06-01T00:00:00Z",
        "updated_at": "2023-06-01T00:00:00Z",
        "size": 36776,
        "download_count": 3
      }
    ]
  }
]=]

---@param base_url string
---@return Result
local function build_result_json(base_url)
  local result_json = result_template_json:format(base_url, base_url, base_url, base_url)
  log.info("DEBUG: using result_json " .. result_json)
  return Result.success(vim.json.decode(result_json))
end

---@type Provider
---@return Result
return {
    github = {
      get_latest_release = function(repo)
        local repo_url = registry_url_template:format(repo)
        return build_result_json(repo_url)
      end,
    }
}
```

When it runs, the result json I built is logged but it does not fetch from those urls.
I configured the download_url_template to start with DOWNLOAD_URL_TEMPLATE for debugging.
I got these logs:

```
[DEBUG Mon 19 Jun 2023 06:23:24 PM EDT] ...l/share/nvim/lazy/mason.nvim/lua/mason-registry/init.lua:48: Failed to scan installation root dir .../.local/share/nvim/lazy/mason.nvim/lua/mason-core/fs.lua:128: ENOENT: no such file or directory: /home/fred/.local/share/nvim/mason/packages
[DEBUG Mon 19 Jun 2023 06:23:24 PM EDT] ...im/lazy/mason.nvim/lua/mason-registry/sources/github.lua:132: Creating registry directory GitHubRegistrySource(repo=mason-org/mason-registry)
[DEBUG Mon 19 Jun 2023 06:23:24 PM EDT] .../.local/share/nvim/lazy/mason.nvim/lua/mason-core/fs.lua:71: fs: mkdirp /home/fred/.local/share/nvim/mason/registries/github/mason-org/mason-registry
[INFO  Mon 19 Jun 2023 06:23:24 PM EDT] /home/fred/.config/nvim/lua/mason-x/provider.lua:42: DEBUG: using result_json   {
    "id": 0,
    "tag_name": "latest",
    "draft": false,
    "prerelease": false,
    "assets": [
      {
        "id": 0,
        "name": "checksums.txt",
        "url": "https://proxy.repo.corporate.com/generic-github-releases/mason-org/mason-registry/releases/latest/download/checksums.txt",
        "browser_download_url": "https://proxy.repo.corporate.com/generic-github-releases/mason-org/mason-registry/releases/latest/download/checksums.txt",
        "created_at": "2023-06-01T00:00:00Z",
        "updated_at": "2023-06-01T00:00:00Z",
        "size": 164,
        "download_count": 4
      },
      {
        "id": 0,
        "name": "registry.json.zip",
        "url": "https://proxy.repo.corporate.com/generic-github-releases/mason-org/mason-registry/releases/latest/download/registry.json.zip",
        "browser_download_url": "https://proxy.repo.corporate.com/generic-github-releases/mason-org/mason-registry/releases/latest/download/registry.json.zip",
        "created_at": "2023-06-01T00:00:00Z",
        "updated_at": "2023-06-01T00:00:00Z",
        "size": 36776,
        "download_count": 3
      }
    ]
  }

[DEBUG Mon 19 Jun 2023 06:23:24 PM EDT] ...ocal/share/nvim/lazy/mason.nvim/lua/mason-core/fetch.lua:34: Fetching URL "https://DOWNLOAD_URL_TEMPLATE.repo.corporate.com/generic-github-releases/mason-org/mason-registry/releases/download/latest/registry.json.zip"
[DEBUG Mon 19 Jun 2023 06:23:24 PM EDT] ...al/share/nvim/lazy/mason.nvim/lua/mason-core/process.lua:116: Spawning cmd="curl", spawn_opts={
  args = { "-H", "User-Agent: mason.nvim (+https://github.com/williamboman/mason.nvim)", "-fsSL", "-X", "GET", "-o", "/home/fred/.local/share/nvim/mason/registries/github/mason-org/mason-registry/registry.json.zip", "https://DOWNLOAD_URL_TEMPLATE.repo.corporate.com/generic-github-releases/mason-org/mason-registry/releases/download/latest/registry.json.zip" }
}
[DEBUG Mon 19 Jun 2023 06:23:24 PM EDT] ...al/share/nvim/lazy/mason.nvim/lua/mason-core/process.lua:162: Spawned with pid 780086
[DEBUG Mon 19 Jun 2023 06:23:24 PM EDT] ...al/share/nvim/lazy/mason.nvim/lua/mason-core/process.lua:148: Job pid=780086 exited with exit_code=6, signal=0
[DEBUG Mon 19 Jun 2023 06:23:24 PM EDT] ...al/share/nvim/lazy/mason.nvim/lua/mason-core/process.lua:116: Spawning cmd="wget", spawn_opts={
  args = { "--header=User-Agent: mason.nvim (+https://github.com/williamboman/mason.nvim)", "-nv", "-o", "/dev/null", "-O", "/home/fred/.local/share/nvim/mason/registries/github/mason-org/mason-registry/registry.json.zip", "--method=GET", "https://DOWNLOAD_URL_TEMPLATE.repo.corporate.com/generic-github-releases/mason-org/mason-registry/releases/download/latest/registry.json.zip" }
}
[DEBUG Mon 19 Jun 2023 06:23:24 PM EDT] ...al/share/nvim/lazy/mason.nvim/lua/mason-core/process.lua:162: Spawned with pid 780090
[DEBUG Mon 19 Jun 2023 06:23:24 PM EDT] ...al/share/nvim/lazy/mason.nvim/lua/mason-core/process.lua:148: Job pid=780090 exited with exit_code=4, signal=0
[ERROR Mon 19 Jun 2023 06:23:24 PM EDT] ...im/lazy/mason.nvim/lua/mason-registry/sources/github.lua:181: Failed to install registry GitHubRegistrySource(repo=mason-org/mason-registry). "Failed to download registry archive."
[ERROR Mon 19 Jun 2023 06:23:24 PM EDT] ...l/share/nvim/lazy/mason.nvim/lua/mason-registry/init.lua:79: Cannot find package "\"gopls\"".
[ERROR Mon 19 Jun 2023 06:23:24 PM EDT] ...l/share/nvim/lazy/mason.nvim/lua/mason-registry/init.lua:79: Cannot find package "\"bash-language-server\"".
[ERROR Mon 19 Jun 2023 06:23:24 PM EDT] ...l/share/nvim/lazy/mason.nvim/lua/mason-registry/init.lua:79: Cannot find package "\"solargraph\"".
[ERROR Mon 19 Jun 2023 06:23:24 PM EDT] ...l/share/nvim/lazy/mason.nvim/lua/mason-registry/init.lua:79: Cannot find package "\"lua-language-server\"".
[ERROR Mon 19 Jun 2023 06:23:24 PM EDT] ...l/share/nvim/lazy/mason.nvim/lua/mason-registry/init.lua:79: Cannot find package "\"json-lsp\"".
```

Mason is configured this way:
```lua
require('mason-x/provider')
require('mason').setup({
  log_level = vim.log.levels.DEBUG,
  providers = { "mason-x.provider", },
  github = {
    download_url_template = "https://DOWNLOAD_URL_TEMPLATE.repo.corporate.com/generic-github-releases/%s/releases/download/%s/%s",
  }
})
```

Do you have any other ideas?

How about this:
Every mason-registry release currently has 2 assets, `registry.json.zip` and `checksums.tzt`.
Add a third asset, `release.json`, which is just the same json that is normally provided by the registry api.
This json provides the release name, urls and size, so if it could be fetched from a known, fixed url then it would solve the problem:

     https://github.com/mason-org/mason-registry/releases/latest/download/release.json

Example release json:
```json
{
  "id": 109148941,
  "tag_name": "2023-06-19-quick-chill",
  "draft": false,
  "prerelease": false,
  "assets": [
    {
      "id": 113452410,
      "url": "https://api.github.com/repos/mason-org/mason-registry/releases/assets/113452410",
      "name": "checksums.txt",
      "browser_download_url": "https://github.com/mason-org/mason-registry/releases/download/2023-06-19-quick-chill/checksums.txt",
      "created_at": "2023-06-19T21:11:47Z",
      "updated_at": "2023-06-19T21:11:47Z",
      "size": 164,
      "download_count": 4
    },
    {
      "id": 113452411,
      "url": "https://api.github.com/repos/mason-org/mason-registry/releases/assets/113452411",
      "name": "registry.json.zip",
      "browser_download_url": "https://github.com/mason-org/mason-registry/releases/download/2023-06-19-quick-chill/registry.json.zip",
      "created_at": "2023-06-19T21:11:47Z",
      "updated_at": "2023-06-19T21:11:47Z",
      "size": 36776,
      "download_count": 3
    }
  ]
}
```

gongfarmer commented:
I am trying to figure out how to get the current release name with an HTTP request that will work inside my corporate mirror.
If I can get that far then the `download_url_template` can be filled in and the rest will just work.

This seems promising:
```
$ curl -L https://proxy.repo.corporatecom/generic-github-releases/mason-org/mason-registry/releases/latest
<!DOCTYPE html>
<html>
<head><meta name="robots" content="noindex" />
<title>Index of generic-github-releases/mason-org/mason-registry/releases/latest</title>
</head>
<body>
<h1>Index of generic-github-releases/mason-org/mason-registry/releases/latest</h1>
<pre>Name                    Last modified      Size</pre><hr/>
<pre><a href="../">../</a>
<a href="download/">download/</a>                18-May-2023 15:22    -
<a href="2023-06-19-frosty-hunt">2023-06-19-frosty-hunt</a>->     -    -
<a href="actions">actions</a>->                    -    -
<a href="collections">collections</a>->                -    -
<a href="enterprise">enterprise</a>->                 -    -
<a href="github-actions">github-actions</a>->             -    -
<a href="issues">issues</a>->                     -    -
<a href="latest">latest</a>->                     -    -
<a href="latest">latest</a>->                     -    -
<a href="mason-org">mason-org</a>->                  -    -
<a href="mason-registry">mason-registry</a>->             -    -
<a href="pricing">pricing</a>->                    -    -
<a href="releases">releases</a>->                   -    -
<a href="security">security</a>->                   -    -
<a href="startups">startups</a>->                   -    -
<a href="topics">topics</a>->                     -    -
<a href="trending">trending</a>->                   -    -
</pre>
<hr/><address style="font-size:small;">Artifactory/7.55.10 Server at proxy.repo.corporate.com Port 80</address></body></html>
```

Hard to parse though.   Too bad I can't get json:

```
$ curl -L https://proxy.repo.corporate.com/generic-github-releases/mason-org/mason-registry/releases/latest -H "Accept: application/json"
{
  "errors" : [ {
    "status" : 406,
    "message" : "Not Acceptable"
  } ]
}
```

I could try writing lua to parse that HTML but that looks like it will be fragile.  I hope you have a better idea.

gongfarmer commented:
I rewrote mason-x/provider.lua to scrape the release name from the HTML listing offered by our internal GitHub mirror server.
This is working well for me.  

Thank you for the guidance.

```lua
-- query repomirror for the name of the latest mason-registry release

local fetch = require "mason-core.fetch"
local log = require "mason-core.log"

---@param str string containing html from repomirror
---@return string
function scrape_release_name(str)
  -- split input string into lines
  lines = {}
  for line in str:gmatch("[^\r\n]+") do
    table.insert(lines, line)
  end

  -- find the index of the first line that contains "download".
  -- @example
  --     <a href="download/">download/</a>                 18-May-2023 15:22    -
  local index = 0
  for k,v in pairs(lines) do
    if string.find(v, "download") then
      index = k
      break
    end
  end

  -- extract release name from the line after the "download" line
  -- @example
  --     <a href="2023-06-20-forked-lenok">2023-06-20-forked-lenok</a>->     -    -
  local line = lines[index + 1]

  -- find indices of quotes surrounding release name
  i_quote1 = string.find(line, '"')
  i_quote2 = string.find(line, '"', i_quote1 + 1)

  return string.sub(line, i_quote1 + 1, i_quote2 - 1)
end

---@type Provider
---@return Result
return {
  github = {
    ---@param repo string (always "mason-org/mason-registry")
    get_latest_release = function(repo)
      local url_template = 'https://proxy.repo.corporate.com/generic-github-releases/%s/releases/latest'
      local result = fetch((url_template):format(repo))
      local release_name = scrape_release_name(result['value'])
      log.fmt_info("mason-x provider found %s release name: '%s'", repo, release_name)

      result['value'] = { tag_name = release_name }
      return result
    end,
  }
}
```

williamboman commented:
Cool, if it works it works :D!

Yeah I was also thinking we could host release metadata in a new asset file which would allow using the `latest` GitHub download URL to gather that information. I'm not a big fan of it however, as I feel like it'd largely be redundant. The "latest" download URL is just a 302 that redirects you to the canonical download URL (with the release version as a URL component), and hosting release metadata that you can only really access if you already have the release metadata (with the exception of the "latest" URL redirection path) feels a bit redundant. There could likely also be additional penalties in terms of slower response times than going through the API.

Finally, I think the installation of the registry serves as a pretty good health check in terms of whether you'll be able to meaningfully use Mason or not. The "providers" mechanism is the escape hatch if you'd like to customize it to your needs, which it seems like you've successfully done.
